### PR TITLE
Update basic00.md

### PR DESCRIPTION
**Pull request by matsduf**
More explicit definition of a valid domain name. Added explicit support for IDN. Added support for underscore requested in issue #450.

### CODE REVIEW

File: docs/specifications/tests/Basic-TP/basic00.md [hunk 1]
```diff
@@ -6,25 +6,45 @@
...
 ### Inputs
 
-The label of the domain name to be tested.
+The domain name to be tested.
 
 ### Ordered description of steps to be taken to execute the test case
 
-1. Remove any trailing dot from the input domain name.
-2. If the total length of the input is more than 253 octets, this test
+1. If the domain name has more than one tailing dot, the test case
```
Comment (pawal) [inline docs/specifications/tests/Basic-TP/basic00.md]: Could be replaced by something that says that any label in the domain name must have a length that is longer than zero octets. Such as the added step 8, irregardless of whether the domain name has been converted from an ULABEL or not.

Comment (matsduf) [inline docs/specifications/tests/Basic-TP/basic00.md]: You are right that step 1 partly covers what step 8 covers. Maybe it is better to change step 1 to "If the input domain name contains two or more consecutive dots, then this test case fails" and remove step 8. Then that simple error will be caught early.

**Comment by sandoche2k** [comment]
@matsduf @vlevigneron @mtoma 

I had a fruitful discussion with @pawal regarding https://github.com/dotse/zonemaster/blob/master/docs/specifications/tests/Basic-TP/basic00.md. We have two issues with it:
1. IDN domains
2. case of underscore : "_sips._tcp.cam.ac.uk"

The case of underscore should be rectified in Syntax

For IDN's, @pawal proposes not to change the current https://github.com/dotse/zonemaster/blob/master/docs/specifications/tests/Basic-TP/basic00.md

Rather, have a separate function such as here : https://godoc.org/github.com/miekg/dns/idn

Hence, from the CLI if we want to test "zonemaster-cli --convert-to-punycode räksmörgås.example"

If the user just tests as "zonemaster-cli räksmörgås.example" without the convert function, he has an advertisement.

But, from the web interface,  we make it implicit. When the web interface encounters non ascii domain, it calls the punycode conversion function in the engine. 

Advantages of this approach: 
The engine core functionality is to test DNS delegation configuration. All other functionalities are outside. 

@bortzmeyer for your view on this approach.

**Comment by bortzmeyer** [comment]
@sandoche2k Seems perfectly fine.

**Comment by matsduf** [comment]
@bortzmeyer and @sandoche2k!

I have a vision of the Zonemaster software using explicit test requirements, Test Cases, to define how the software behave when domains are tested. For me, the Zonemaster software consists of the Engine, the CLI, the Backend and the GUI. My vision is that the Test Cases covers all parts of Zonemaster.

We have Test Case BASIC00 that restricts the allowable characters in the domain name to be tested and also put some lenght restrictions on the domain name. BASIC00 is only partly implemented in Engine (length restrictions), and that is inherited in CLI and Backend. GUI has its own implementation, which seems to follow BASIC00 with the exception of IDN.

Even though BASIC00 forbids IDN names as U-labels to be tested, CLI, Backend and GUI have implemented support for U-labels.

What is your suggestion for BASIC00? What characters should Zonemaster support? How do we test that? We could allow a larger set of characters than the traditional host name, but then we should know that it works. Where are the real (not theoretical) use cases for domain names beyond the set that I suggested?

Remember that BASIC00 is for Zonemaster, not just for the Engine.

**Comment by pawal** [comment]
This is not how we wrote the specifications from the start. The specifications only cover the behaviour of the engine, and it has been continuing to do so so far.

Using the engine there are some applications, and applications in the core project is the CLI and the GUI. There are applications outside of the core project such as several different programs I wrote that is only using the engine - as it was designed, using the exposed Perl APIs. It would be sad to see a direction where the separation of the testing library and the core applications are mixed up, thus making it harder to write software using the test library.

There can still be applications that limits the scope of what should be tested, such as some TLDs that only allow the testing of their own TLD, or a completely different set of restrictions, in the same way that you can write your own custom policy. Limiting those things in the engine is putting restrictions on all applications, and that would be very unfortunate.

That said, I would advise to separate specifications in the core project from the engine and the applications. The applications should have their own specifications, not only because they are different in scope, but also to not limit others to write applications for other uses.

**Comment by matsduf** [comment]
@pawal, BASIC00 is there and has not been fully implemented in the Engine. Most people that go to the Zonemaster repository and find the specifications will expect that they are applied for all parts of Zonemaster, included CLI, Backend and GUI.

I still want to see real use case where characters beyond what I suggested are allowed. And the character set code be expanded, but it must be tested and correctly documented.

**Comment by pawal** [comment]
It is very simple, to test delegations such as these (or OTHER yet unknown variants):

```
$~>dig ns räksmörgås.tset.se @vic20.blipp.com

; <<>> DiG 9.10.3-P4-Debian <<>> ns räksmörgås.tset.se @vic20.blipp.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 62815
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 2, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;r\195\164ksm\195\182rg\195\165s.tset.se. IN NS

;; AUTHORITY SECTION:
r\195\164ksm\195\182rg\195\165s.tset.se. 86400 IN NS ns3.nic.se.
r\195\164ksm\195\182rg\195\165s.tset.se. 86400 IN NS ns.nic.se.

;; Query time: 4 msec
;; SERVER: 2001:16d8:ff00:2a9::2#53(2001:16d8:ff00:2a9::2)
;; WHEN: Fri Nov 25 19:11:24 CET 2016
;; MSG SIZE  rcvd: 89
```

**Comment by pawal** [comment]
This is fine if it is possible to disable basic00 in any test that I want to run.

**Comment by sandoche2k** [comment]
@matsduf As there seems to be a consensus, i think you can merge the updates for basic00. @vlevigneron is waiting for spec updates to start coding.

**Comment by pawal** [comment]
There is no way to disable basic00, or did you fix this also? Otherwise, please revert this to previous behavior.

**Pull request merged**